src/zope/schema/_field.py

```diff
@@ -379,19 +394,34 @@ def __init__(self, values=None, vocabulary=None, source=None, **kw):
 
     source = property(lambda self: self.vocabulary)
 
-    def bind(self, object):
-        """See zope.schema._bootstrapinterfaces.IField."""
-        clone = super(Choice, self).bind(object)
-        # get registered vocabulary if needed:
-        if IContextSourceBinder.providedBy(self.vocabulary):
-            clone.vocabulary = self.vocabulary(object)
-        elif clone.vocabulary is None and self.vocabularyName is not None:
+    def _resolve_vocabulary(self, value):
+        # Find the vocabulary we should use, raising
+        # an exception if this isn't possible, and returning
+        # an ISource otherwise.
+        vocabulary = self.vocabulary
+        if IContextSourceBinder.providedBy(vocabulary) and self.context is not None:
+            vocabulary = vocabulary(self.context)
+        elif vocabulary is None and self.vocabularyName is not None:
             vr = getVocabularyRegistry()
-            clone.vocabulary = vr.get(object, self.vocabularyName)
+            try:
+                vocabulary = vr.get(self.context, self.vocabularyName)
+            except LookupError:
+                raise MissingVocabularyError(
+                    "Can't validate value without vocabulary named %r" % (self.vocabularyName,)
+                ).with_field_and_value(self, value)
+
+        if not ISource.providedBy(vocabulary):
+            raise InvalidVocabularyError(vocabulary).with_field_and_value(self, value)
 
-        if not ISource.providedBy(clone.vocabulary):
-            raise ValueError('Invalid clone vocabulary')
+        return vocabulary
 
+    def bind(self, context):
+        """See zope.schema._bootstrapinterfaces.IField."""
+        clone = super(Choice, self).bind(context)
+        # Eagerly get registered vocabulary if needed;
+        # once that's done, just return it
+        clone.vocabulary = clone._resolve_vocabulary(None)
+        clone._resolve_vocabulary = lambda value: clone.vocabulary
```

Comment:
Why shadow the `_resolve_vocabulary` method here?  Wouldn't it be better to make `_resolve_vocabulary` return `self.vocabulary` if it is set?

Reply:
`self.vocabulary` is over-loaded to sometimes hold a `IContextSourceBinder`, and not an actual `ISource` (let alone `IVocabulary`). (This is weird, IMO.) 

The shadowing here is to avoid the relative expense of checking `ISource.providedBy` and the various other checks that `_resolve_vocabulary` does (I've recently been profiling an app that makes heavy use of schema fields and interfaces, and `IFoo.providedBy` for various `IFoo` showed up as a hotspot that it paid to avoid). Assuming you're binding the field to validate something with it, it's guaranteed to be called again at least once (from `_validate`), and all of that work would be duplicated/wasted.